contracts/Template.sol

```diff
@@ -1,106 +1,159 @@
+        bytes memory initializeData = abi.encodeWithSelector(CounterApp(0).initialize.selector);
+        CounterApp counterApp = CounterApp(_installDefaultApp(_dao, _appId, initializeData));
+        counterApp.initialize();
+        return counterApp;
+    }
 
-        acl.createPermission(this, tokenManager, tokenManager.MINT_ROLE(), this);
-        tokenManager.mint(root, 1); // Give one token to root
+    function _setupPermissions(
+        Kernel _dao,
+        ACL _acl,
+        Voting _voting,
+        TokenManager _tokenManager,
+        CounterApp _app
+    )
+        internal 
+    {
+        _createEvmScriptsRegistryPermissions(_acl, _voting, _voting);
+        _createCustomVotingPermissions(_acl, _voting, _tokenManager);
+        _createCustomTokenManagerPermissions(_acl, _tokenManager, _voting);
+        // Set app permission
+        _createCustomAppPermissions(_acl, _app, _voting);
+        _transferRootPermissionsFromTemplateAndFinalizeDAO(_dao, _voting);
+    }
 
-        acl.createPermission(ANY_ENTITY, voting, voting.CREATE_VOTES_ROLE(), root);
+    function _createCustomVotingPermissions(
+        ACL _acl,
+        Voting _voting,
+        TokenManager _tokenManager
+    ) 
+        internal
+    {
+        _acl.createPermission(_tokenManager, _voting, _voting.CREATE_VOTES_ROLE(), _voting);
+        _acl.createPermission(_voting, _voting, _voting.MODIFY_QUORUM_ROLE(), _voting);
+        _acl.createPermission(_voting, _voting, _voting.MODIFY_SUPPORT_ROLE(), _voting);
+    }
 
-        acl.createPermission(voting, app, app.INCREMENT_ROLE(), voting);
-        acl.createPermission(ANY_ENTITY, app, app.DECREMENT_ROLE(), root);
-        acl.grantPermission(voting, tokenManager, tokenManager.MINT_ROLE());
+    function _createCustomTokenManagerPermissions(
+        ACL _acl,
+        TokenManager _tokenManager,
+        Voting _voting
+    ) 
+        internal 
+    {
+        _acl.createPermission(_voting, _tokenManager, _tokenManager.BURN_ROLE(), _voting);
+        _acl.createPermission(_voting, _tokenManager, _tokenManager.MINT_ROLE(), _voting);
+    }
 
-        // Clean up permissions
-        acl.grantPermission(root, dao, dao.APP_MANAGER_ROLE());
-        acl.revokePermission(this, dao, dao.APP_MANAGER_ROLE());
-        acl.setPermissionManager(root, dao, dao.APP_MANAGER_ROLE());
+    function _createCustomAppPermissions(
+        ACL _acl,
+        CounterApp _app,
+        Voting _voting
+    ) 
+        internal
+    {
+        _acl.createPermission(_voting, _app, _app.INCREMENT_ROLE(), _voting);
+        _acl.createPermission(ANY_ENTITY, _app, _app.DECREMENT_ROLE(), _voting);
+    }
 
-        acl.grantPermission(root, acl, acl.CREATE_PERMISSIONS_ROLE());
-        acl.revokePermission(this, acl, acl.CREATE_PERMISSIONS_ROLE());
-        acl.setPermissionManager(root, acl, acl.CREATE_PERMISSIONS_ROLE());
+    function _cacheToken(
```

Comment:
We now have the [`TokenCache`](https://github.com/aragon/dao-templates/blob/master/shared/contracts/TokenCache.sol) utility available, so we can remove these parts :).